modules/460-log-shipper/hooks/generate_config.go

```diff
@@ -181,6 +205,99 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	ClientKey  []byte
+	KeyPass    []byte
+}
+
+func getTLSConfigFromSecretSnapshot(name string, snapshot []go_hook.FilterResult) (*tlsConfigFromSecret, error) {
+	for _, secret := range snapshot {
+		if secret.(*corev1.Secret).Name == name {
+			s := secret.(*corev1.Secret)
+			return &tlsConfigFromSecret{
+				CACert:     s.Data["ca.pem"],
+				ClientCert: s.Data["crt.pem"],
+				ClientKey:  s.Data["key.pem"],
+				KeyPass:    s.Data["keyPass"],
+			}, nil
+		}
+	}
+	return nil, fmt.Errorf("secret %s not found", name)
+}
+
+func getTLSSecretRef(dest v1alpha1.ClusterLogDestination) string {
+	switch dest.Spec.Type {
+	case "Elasticsearch":
+		if dest.Spec.Elasticsearch.TLS.SecretRef != nil {
+			return dest.Spec.Elasticsearch.TLS.SecretRef.Name
+		}
+	case "Vector":
+		if dest.Spec.Vector.TLS.SecretRef != nil {
+			return dest.Spec.Vector.TLS.SecretRef.Name
+		}
+	case "Loki":
+		if dest.Spec.Loki.TLS.SecretRef != nil {
+			return dest.Spec.Loki.TLS.SecretRef.Name
+		}
+	case "Splunk":
+		if dest.Spec.Splunk.TLS.SecretRef != nil {
+			return dest.Spec.Splunk.TLS.SecretRef.Name
+		}
+	case "Logstash":
+		if dest.Spec.Logstash.TLS.SecretRef != nil {
+			return dest.Spec.Logstash.TLS.SecretRef.Name
+		}
+	case "Socket":
+		if dest.Spec.Socket.TCP.TLS.SecretRef != nil {
+			return dest.Spec.Socket.TCP.TLS.SecretRef.Name
+		}
+	case "Kafka":
+		if dest.Spec.Kafka.TLS.SecretRef != nil {
+			return dest.Spec.Kafka.TLS.SecretRef.Name
+		}
+	}
+	return ""
+}
+
+func applyTLSConfig(dest *v1alpha1.ClusterLogDestination, tlsConfig *tlsConfigFromSecret) error {
+	accessors := map[string]func() *v1alpha1.CommonTLSSpec{
+		"Elasticsearch": func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Elasticsearch.TLS },
+		"Vector":        func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Vector.TLS },
+		"Loki":          func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Loki.TLS },
+		"Splunk":        func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Splunk.TLS },
+		"Logstash":      func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Logstash.TLS },
+		"Socket":        func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Socket.TCP.TLS },
+		"Kafka":         func() *v1alpha1.CommonTLSSpec { return &dest.Spec.Kafka.TLS },
+	}
+
+	getTLSSpec := accessors[dest.Spec.Type]
+	if getTLSSpec == nil {
+		return errors.Errorf("no tls spec accessor for destination type %s", dest.Spec.Type)
+	}
+
+	tls := getTLSSpec()
+	if tls.SecretRef == nil {
+		return nil
+	}
+
+	if len(tlsConfig.CACert) > 0 {
+		tls.CAFile = base64.RawStdEncoding.EncodeToString(tlsConfig.CACert)
+	}
+	if len(tlsConfig.ClientCert) > 0 {
+		tls.CommonTLSClientCert.CertFile = base64.StdEncoding.EncodeToString(tlsConfig.ClientCert)
+	}
+	if len(tlsConfig.ClientKey) > 0 {
+		tls.CommonTLSClientCert.KeyFile = base64.StdEncoding.EncodeToString(tlsConfig.ClientKey)
+	}
+	if len(tlsConfig.KeyPass) > 0 {
+		tls.CommonTLSClientCert.KeyPass = base64.StdEncoding.EncodeToString(tlsConfig.KeyPass)
+	}
+
+	return nil
+}
```

Comment:
I can't say I like this approach mainly because we check the destination type twice.

Instead, I suggest a different approach, so the functions above can be replaced with
```suggestion
func getTLSConfig(dest v1alpha1.ClusterLogDestination) *v1alpha1.CommonTLSSpec {
	switch dest.Spec.Type {
	case "Elasticsearch":
		return &dest.Spec.Elasticsearch.TLS
	case "Vector":
		return &dest.Spec.Vector.TLS
	case "Loki":
		return &dest.Spec.Loki.TLS
	case "Splunk":
		return &dest.Spec.Splunk.TLS
	case "Logstash":
		return &dest.Spec.Logstash.TLS
	case "Socket":
		return &dest.Spec.Socket.TCP.TLS
	case "Kafka":
		return &dest.Spec.Kafka.TLS
	}

	return nil
}

func applyTLSConfig(destTlsConfig *v1alpha1.CommonTLSSpec, tlsConfig *tlsConfigFromSecret) {
	if len(tlsConfig.CACert) > 0 {
		destTlsConfig.CAFile = base64.RawStdEncoding.EncodeToString(tlsConfig.CACert)
	}
	if len(tlsConfig.ClientCert) > 0 {
		destTlsConfig.CommonTLSClientCert.CertFile = base64.StdEncoding.EncodeToString(tlsConfig.ClientCert)
	}
	if len(tlsConfig.ClientKey) > 0 {
		destTlsConfig.CommonTLSClientCert.KeyFile = base64.StdEncoding.EncodeToString(tlsConfig.ClientKey)
	}
	if len(tlsConfig.KeyPass) > 0 {
		destTlsConfig.CommonTLSClientCert.KeyPass = base64.StdEncoding.EncodeToString(tlsConfig.KeyPass)
	}
}
```
So the `getTLSConfig` is the only one responsible for selecting the proper TLS config to patch and the `getTLSConfig` just updates necessary fields.

And we have to update the caller code as well. See below